init.go

```diff
@@ -33,84 +42,74 @@ func opensslInit(file string) error {
 			_OPENSSL_INIT_LOAD_CONFIG|
 			_OPENSSL_INIT_LOAD_CRYPTO_STRINGS,
 		nil); err != nil {
-		dlclose(handle)
+		close()
 		return err
 	}
 	osslHandle = handle
 	return nil
 }
 
-// loadFuncs loads and initialize the OpenSSL functions.
-// See shims.go for the complete list of supported functions.
-func loadOpenSSLFuncs(handle unsafe.Pointer) {
-	mkcgoLoad_(handle)
-	if vMajor == 1 {
-		mkcgoLoad_legacy_1(handle)
-		if vPatch == 1 {
-			mkcgoLoad_111(handle)
-		}
-	} else {
-		mkcgoLoad_111(handle)
-		mkcgoLoad_3(handle)
-	}
-}
-
 // initForCheckVersion loads and initialize only the
 // functions required in [CheckVersion].
 // It returns a function that must be called to release the resources.
 // The function leaves all the global variables in the same state as they were
 // before the call.
 func initForCheckVersion(file string) (func(), error) {
-	// This function can be called when the
 	prevMajor, prevMinor, prevPatch := vMajor, vMinor, vPatch
-	// Load the OpenSSL shared library using dlopen.
-	handle, err := openLibrary(file)
+	handle, close, err := openLibrary(file)
 	if err != nil {
 		vMajor, vMinor, vPatch = prevMajor, prevMinor, prevPatch
 		return nil, err
 	}
-
+	var loadX func(unsafe.Pointer)
+	var unloadX func()
 	switch vMajor {
 	case 1:
-		mkcgoLoad_init_1(handle)
+		loadX = mkcgoLoad_init_1
+		unloadX = mkcgoUnload_init_1
 	case 3:
-		mkcgoLoad_init_3(handle)
+		loadX = mkcgoLoad_init_3
+		unloadX = mkcgoUnload_init_3
 	default:
 		// We shouldn't get here: openLibrary should have already returned an error.
 		panic(errUnsupportedVersion())
 	}
+	loadX(handle)
 	return func() {
-		dlclose(handle)
-		// Undo all the changes made in this function.
-		mkcgoUnload_version()
-		mkcgoUnload_init_1()
-		mkcgoUnload_init_3()
-		vMajor, vMinor, vPatch = prevMajor, prevMinor, prevPatch
+		close()
 		if osslHandle != nil {
-			loadOpenSSLFuncs(osslHandle)
+			loadX(osslHandle)
```

Comment:
Is this call being done (along with the version reset) to make it so that you can call `opensslInit` *then* `initForCheckVersion`, and `initForCheckVersion` will restore the state?

Is it a problem if we have OpenSSL 1 loaded, then call `initForCheckVersion` on OpenSSL 3, and then we end up calling `mkcgoLoad_init_3` on an OpenSSL 1 handle? I think that would result in an error for the non-optional bits.

Reply:
d'oh, good catch.